### ISSUE TITLE
No apply function found matching unapply return types

### ISSUE THREAD

**Issue by alexbaretta** [playframework/playframework#4677]
I see that similar issues have been reported in the past. I would like to report a very simple case that triggers this bug. I'm using Play 2.3.7 and Scala 2.11.2.

In the following snippet I show that Json.reads works with a case class with a single Int parameter, fails with a single Array[Int] parameter, works with a single List[Int] and works agains with two params, Array[Int] and Int. This seems to imply that the bug is triggered specifically when a case class has a single parameter of type Array[...].

Welcome to Scala version 2.11.2 (Java HotSpot(TM) 64-Bit Server VM, Java 1.7.0_71).
Type in expressions to have them evaluated.
Type :help for more information.

scala> import play.api.libs.json._
import play.api.libs.json._

scala> case class X(x : Int)
defined class X

scala> Json.reads[X]
res0: play.api.libs.json.Reads[X] = play.api.libs.json.Reads$$anon$8@37bb9822

scala> case class X(x : Array[Int])
defined class X

scala> Json.reads[X]
<console>:13: error: No apply function found matching unapply return types
              Json.reads[X]
                        ^

scala> case class X(x : List[Int])
defined class X

scala> Json.reads[X]
res2: play.api.libs.json.Reads[X] = play.api.libs.json.Reads$$anon$8@2d09aa60

scala> case class X(x : Array[Int], xx : Int)
defined class X

scala> Json.reads[X]
res3: play.api.libs.json.Reads[X] = play.api.libs.json.Reads$$anon$8@544a7e3e

**Comment by jroper** [comment]
I wonder if it has something to do with Java varargs.

**Comment by jroper** [comment]
Wait, it looks like this is fixed in Play 2.4:

```scala
scala> case class X(x: Array[Int])
defined class X

scala> Json.reads[X]
res6: play.api.libs.json.Reads[X] = play.api.libs.json.Reads$$anon$7@57d13f38
```

**Comment by jroper** [comment]
I wonder if this is related to #3587.  Anyway, closing because it seems to be fixed in 2.4.
